File: app/src/main/java/au/id/blackwell/kurt/lantv/utility/LimitedPool.java
package au.id.blackwell.kurt.lantv.utility;

import java.util.ArrayDeque;
import java.util.ArrayList;

public class LimitedPool<T> implements Pool<T> {
    private final ArrayList<T> mItems = new ArrayList<T>();
    private final ArrayDeque<Callback> mQueue = new ArrayDeque<Callback>();

    public void addItem(T item) {
        mItems.add(item);
    }

    public void removeItem(T item) {
        mItems.remove(item);
    }

    @Override
    public void request(Callback callback) {
        mQueue.add(callback);
        update();
    }

    @Override
    public void cancel(Callback callback) {
        mQueue.remove(callback);
    }

    private void update() {
        if (!mItems.isEmpty() && !mQueue.isEmpty()) {
            final Callback requester = mQueue.pop();
            final T item = mItems.remove(mItems.size() - 1);

            requester.run(new Item() {
                @Override
                public T get() {
                    return item;
                }

                @Override
                public void release() {
                    addItem(item);
                }
            });
        }
    }
}

File: app/src/main/java/au/id/blackwell/kurt/lantv/utility/LimitedPool.kt
package au.id.blackwell.kurt.lantv.utility

import io.reactivex.Observable
import io.reactivex.subjects.SingleSubject
import java.util.ArrayDeque
import java.util.ArrayList

class LimitedPool<T> : Pool<T> {
    private val mItems = ArrayList<T>()
    private val mQueue = ArrayDeque<SingleSubject<T>>()

    override fun addItem(item: T) {
        mItems.add(item)
    }

    override fun removeItem(item: T) {
        mItems.remove(item)
    }

    override fun request(): Observable<T> {
        val subject = SingleSubject.create<T>()
        mQueue.add(subject)
        update()

        return subject.toObservable()
            .doOnDispose({ mQueue.remove(subject) })
    }

    private fun update() {
        if (!mItems.isEmpty() && !mQueue.isEmpty()) {
            val subject = mQueue.pop()
            val item = mItems.removeAt(mItems.size - 1)

            subject.onSuccess(item)
        }
    }
}
